### PR TITLE
refactor(snapshot): simplify snapshot failure propagation & handling

### LINKED ISSUE #7239: Refactor the ReceivedSnapshot#apply interface and implementations
**Description**

There was some confusion in a recent PR over how errors should be handled in FileBasedReceivedSnapshot#apply since there were several options. The apply method for the ReceivedSnapshot interface looks like:
`ActorFuture<Boolean> apply(SnapshotChunk chunk) throws IOException`

At first glance there's three ways failures can be handled: throwing an IOException, completing the future exceptionally and completing it with false. This is more confusing since I think FileBasedReceivedSnapshot wouldn't let an exception bubble up instead failing the actor. 

We could refactor this to uniformly handle failures in one way. [My suggestion](https://github.com/camunda-cloud/zeebe/pull/7204#discussion_r648373515) was to complete the future exceptionally. It would remove the IOException and also switch the `Future<Boolean>` with a `Future<Void>`. Consumers would then only check how the future was completed and log/recover accordingly.

### PR DESCRIPTION
**Pull request by miguelpires**
## Description

The ReceivedSnapshot's apply method looks like: `ActorFuture<Boolean> apply(SnapshotChunk chunk) throws IOException` . This means that there are three ways of handling failures: by completing the future with false, completing it exceptionally or by throwing an IOException. Having several ways of propagating failures makes implementations more confusing and also forces consumers to handle all of them to be safe. This PR simplifies the interface to only propagate in one way which, in turn, simplifies its implementations and consumers.

## Related issues

closes #7239 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:243]: :+1:

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java
Comment (npepinpe) [inline snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java:65]: Why not just used `actor.submit`? I mean, we don't care about a return value, so call is probably not necessary.

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java
Comment (npepinpe) [inline snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java:161]: Actually, the second condition isn't about format but it's simply the wrong ID, so in that sense the message is confusing. Consider splitting these - it's more verbose, but it saves time when diagnosing issues.

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java
Comment (npepinpe) [inline snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java:170]: Out of scope, but since we're here and you wrote this: would it make sense to use `CREATE, WRITE, TRUNCATE` instead? `CREATE_NEW` will cause this to fail if the file already exists. Would we care if we were overwriting it? Probably not? Anyhow, it probably doesn't happen, but it was just a thought.

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/SnapshotWriteException.java
Comment (npepinpe) [inline snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/SnapshotWriteException.java:10]: :tada:

**Review by npepinpe — APPROVED** [review]
:rocket: 

Great stuff! I had one small comment about confusing message and one about using `actor.submit` instead of `call`, but I don't see any blockers which would require a second review pass.

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java
Comment (MiguelPires) [reply snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java:65]: Submit takes a Runnable which wouldn't allow `applyInternal` to throw an exception. That's also why there's a seemingly useless `return null` there. It's there to cast the lambda to Callable instead of Runnable

Comment (npepinpe) [reply snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java:65]: Ah, I see, you want to be able to throw. We have the concept of a `CheckedRunnable`, but I don't think we use it in the actor scheduler. I think that'd be more than a little bit of refactoring, so let's go with call. :+1:

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java
Comment (MiguelPires) [reply snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java:170]: I'm not sure if that situation can actually occur but I think anyway it would be an easy change so we could do that

**Pull request merged**
